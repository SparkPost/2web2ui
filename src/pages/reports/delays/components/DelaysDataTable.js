import React, { Component } from 'react';
import _ from 'lodash';
import styled from 'styled-components';
import { Panel, Table } from 'src/components/matchbox';
import { TableCollection, Empty, LongTextContainer } from 'src/components';
import { Heading } from 'src/components/text';
import { Percent } from 'src/components/formatters';
import { safeRate } from 'src/helpers/math';
import AddFilterLink from '../../components/AddFilterLink';

const NoPadding = styled.div`
  padding: 0;
`;

const columns = [
  { label: 'Reason', sortKey: 'reason', width: '45%' },
  { label: 'Domain', sortKey: 'domain' },
  { label: 'Total Delays', sortKey: 'count_delayed' },
  { label: 'Delayed First Attempt (%)', sortKey: 'count_delayed_first' },
];

export class DelaysDataTable extends Component {
  getRowData = rowData => {
    const { totalAccepted } = this.props;
    const { reason, domain, count_delayed, count_delayed_first } = rowData;
    return [
      <LongTextContainer text={reason} />,
      <AddFilterLink
        newFilter={{ type: 'Recipient Domain', value: domain }}
        reportType="delayed"
        content={domain}
      />,
      count_delayed,
      <span>
        {count_delayed_first} (<Percent value={safeRate(count_delayed_first, totalAccepted)} />)
      </span>,
    ];
  };

  TableWrapper = props => (
    <Panel.LEGACY>
      <Panel.LEGACY.Section>
        <Heading as="h3" looksLike="h4">
          Delayed Messages
        </Heading>
      </Panel.LEGACY.Section>

      <NoPadding as={Panel.LEGACY.Section}>
        <Table>{props.children}</Table>
      </NoPadding>
    </Panel.LEGACY>
  );

  render() {
    const { rows, title } = this.props;

    if (_.isEmpty(rows)) {
      return (
        <Panel.LEGACY>
          <Empty message="No delayed messages to report" />
        </Panel.LEGACY>
      );
    }

    return (
      <TableCollection
        columns={columns}
        rows={rows}
        getRowData={this.getRowData}
        pagination
        defaultSortColumn="count_delayed"
        defaultSortDirection="desc"
        title={title}
        wrapperComponent={this.TableWrapper}
      />
    );
  }
}

export default DelaysDataTable;
