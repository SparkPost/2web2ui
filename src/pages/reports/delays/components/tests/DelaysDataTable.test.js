import React from 'react';
import { DelaysDataTable } from '../DelaysDataTable';
import { shallow } from 'enzyme';

describe('DelaysDataTable: ', () => {
  let wrapper;

  const props = {
    rows: [
      {
        count_delayed: 100,
        count_delayed_first: 1,
        reason: 'my reason'
      }
    ],
    totalAccepted: 50,
    addFilters: jest.fn()
  };

  beforeEach(() => {
    wrapper = shallow(<DelaysDataTable {...props} />);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });


  it('should render page', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should show empty message when there are no reasons', () => {
    wrapper.setProps({ rows: null });
    expect(wrapper).toMatchSnapshot();
    wrapper.setProps({ rows: []});
    expect(wrapper).toMatchSnapshot();
  });

  it('should render row data properly', () => {
    const rows = wrapper.instance().getRowData({ reason: 'bad delay', count_delayed: 1, count_delayed_first: 10, domain: 'gmail.com' });
    expect(rows).toMatchSnapshot();
  });

});
