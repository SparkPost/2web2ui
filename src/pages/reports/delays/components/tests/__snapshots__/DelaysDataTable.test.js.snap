// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DelaysDataTable:  should render page 1`] = `
<TableCollection
  columns={
    Array [
      Object {
        "label": "Reason",
        "sortKey": "reason",
        "width": "45%",
      },
      Object {
        "label": "Domain",
        "sortKey": "domain",
      },
      Object {
        "label": "Total Delays",
        "sortKey": "count_delayed",
      },
      Object {
        "label": "Delayed First Attempt (%)",
        "sortKey": "count_delayed_first",
      },
    ]
  }
  defaultSortColumn="count_delayed"
  defaultSortDirection="desc"
  getRowData={[Function]}
  pagination={true}
  rows={
    Array [
      Object {
        "count_delayed": 100,
        "count_delayed_first": 1,
        "reason": "my reason",
      },
    ]
  }
  wrapperComponent={[Function]}
/>
`;

exports[`DelaysDataTable:  should render row data properly 1`] = `
Array [
  <LongTextContainer
    text="bad delay"
  />,
  <Memo(Connect(AddFilterLink))
    content="gmail.com"
    newFilter={
      Object {
        "type": "Recipient Domain",
        "value": "gmail.com",
      }
    }
    reportType="delayed"
  />,
  1,
  <span>
    10
     (
    <Percent
      value={20}
    />
    )
  </span>,
]
`;

exports[`DelaysDataTable:  should show empty message when there are no reasons 1`] = `
<Panel.LEGACY>
  <Empty
    message="No delayed messages to report"
  />
</Panel.LEGACY>
`;

exports[`DelaysDataTable:  should show empty message when there are no reasons 2`] = `
<Panel.LEGACY>
  <Empty
    message="No delayed messages to report"
  />
</Panel.LEGACY>
`;
